R/data.table.R

```diff
@@ -1881,17 +1881,58 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
 #    x
 #}
 
-
-as.matrix.data.table <- function(x,...)
-{
-  dm <- dim(x)
-  cn <- names(x)
+as.matrix.data.table <- function(x, rownames, ...) {
+  rn <- NULL
+  rnc <- NULL
+  if (!missing(rownames)) { # Convert rownames to a column index if possible
+    if (is.null(rownames)) {
+      warning("rownames is NULL, ignoring rownames")
+    } else if (length(rownames) != 1) {
+      stop("rownames must be a single column in x")
+    } else if (is.na(rownames)) {
+      warning("rownames is NA, ignoring rownames")
+    } else if (is.logical(rownames) && !isTRUE(rownames)) {
+      warning("rownames is FALSE, ignoring rownames")
+    } else if (!(is.logical(rownames) || is.character(rownames) || is.numeric(rownames))) {
+      # E.g. because rownames is some sort of object that cant be converted to a column index
+      stop("rownames must be TRUE, a column index, or a column name in x")
+    } else {
+      if (is.logical(rownames) && isTRUE(rownames)) {
+        if (haskey(x)) { 
+          rownames <- key(x) 
+          if (length(rownames) > 1) {
+            warning("rownames is TRUE but multiple keys found in key(x), using first column instead")
+            rownames <- 1
+          }
+        } else { 
+          rownames <- 1
+        }
+      }
+      if (is.character(rownames)) { # Handles cases where rownames is a column name, or key(x) from TRUE
+        rnc <- chmatch(rownames, names(x))
+        if (is.na(rnc)) stop(rownames, " is not a column of x")
+      } else { # rownames is an index already
+        if (rownames < 1 || rownames > ncol(x))
+          stop("rownames is ", rownames, " which is outside the column number range [1,ncol=", ncol(x), "]")
+        rnc <- rownames
+      }
+    }
+  }
+  if (!is.null(rnc)) { # If there are rownames, extract and drop that column
+    rn <- x[[rnc]]
+    dm <- dim(x) - c(0, 1)
+    cn <- names(x)[-rnc]
+    X <- x[, -rnc, with = FALSE]
```

Comment:
I think `x[, .SD, .SDcols = c(cn)]` or `x[, (rn) := NULL]` could work -- not 100% on what the variables are, or the best way to do this *within* data.table.